// @zappar/zappar/lib/facelandmark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zappar_1 = require("./zappar");
const gl_matrix_1 = require("gl-matrix");
var zappar_cv_1 = require("@zappar/zappar-cv");
exports.FaceLandmarkName = zappar_cv_1.face_landmark_name_t;
class FaceLandmark {
    constructor(_name) {
        this._name = _name;
        this.pose = gl_matrix_1.mat4.create();
        this._z = zappar_1.z();
        this._impl = this._z.face_landmark_create(this._name);
    }
    destroy() {
        this._z.face_landmark_destroy(this._impl);
    }
    updateFromFaceAnchor(f, mirror) {
        this._z.face_landmark_update(this._impl, f.identity, f.expression, mirror || false);
        this.pose = this._z.face_landmark_anchor_pose(this._impl);
    }
    updateFromIdentityExpression(identity, expression, mirror) {
        this._z.face_landmark_update(this._impl, identity, expression, mirror || false);
        this.pose = this._z.face_landmark_anchor_pose(this._impl);
    }
    _getImpl() {
        return this._impl;
    }
}
exports.FaceLandmark = FaceLandmark;
