// @zappar/zappar/lib/zappar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zappar_cv_1 = require("@zappar/zappar-cv");
const gl_matrix_1 = require("gl-matrix");
const version_1 = require("./version");
let _z;
function z() {
    if (!_z) {
        console.log(`Zappar JS v${version_1.VERSION}`);
        _z = zappar_cv_1.initialize();
    }
    return _z;
}
exports.z = z;
function cameraDefaultDeviceID(userFacing) {
    return z().camera_default_device_id(userFacing || false);
}
exports.cameraDefaultDeviceID = cameraDefaultDeviceID;
function invert(m) {
    let ret = gl_matrix_1.mat4.create();
    gl_matrix_1.mat4.invert(ret, m);
    return ret;
}
exports.invert = invert;
function projectionMatrixFromCameraModel(model, renderWidth, renderHeight) {
    return z().projection_matrix_from_camera_model(model, renderWidth, renderHeight);
}
exports.projectionMatrixFromCameraModel = projectionMatrixFromCameraModel;
function drawPlane(gl, projectionMatrix, cameraMatrix, targetMatrix, texture) {
    z().draw_plane(gl, projectionMatrix, cameraMatrix, targetMatrix, texture);
}
exports.drawPlane = drawPlane;
function browserIncompatible() {
    return z().browser_incompatible();
}
exports.browserIncompatible = browserIncompatible;
function browserIncompatibleUI() {
    z().browser_incompatible_ui();
}
exports.browserIncompatibleUI = browserIncompatibleUI;
